### PR TITLE
feat: add common timeouts to registry/digest/head fetches

### PR DESCRIPTION
**Pull request by piksel**
This PR moves all the registry communications into methods on a common `Client` struct and adds a way to control and enforce timeouts during it's operation.

Previously, all requests used their own http.Client with only one of them being configured for timeout. The user experience when a registry didn't respond to a request was that watchtower just froze permanently during a session, with no output. See #1490.

This also makes it possible to fully test the registry comms without connecting to an actual registry server, but only the bare minimum required for testing the timeout has been added to this PR.

### CODE REVIEW

File: pkg/registry/client.go [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+		DialContext: (&net.Dialer{
+			Timeout:   30 * time.Second,
+			KeepAlive: 30 * time.Second,
+		}).DialContext,
+		ForceAttemptHTTP2:     true,
+		MaxIdleConns:          100,
+		IdleConnTimeout:       90 * time.Second,
+		TLSHandshakeTimeout:   10 * time.Second,
+		ExpectContinueTimeout: 1 * time.Second,
+		TLSClientConfig:       &tls.Config{InsecureSkipVerify: true},
```
[the remark this answers is not in the document]
Comment (piksel) [reply pkg/registry/client.go]: This was just moved from another file, but I don't know why we would want this. The usual reason is for talking to local servers, but the HEAD requests are mostly useful for limiting the traffic to rate limited registries.

@simskij do you remember the justification?

**Comment by trogo** [comment]
Requesting this PR to be merged because this issue happens quite frequently in our environment.

**Review by 0220tiezo1323 — CHANGES_REQUESTED** [review]
No
